### PR TITLE
[SC-164] Add Polygon zkEVM

### PR DESCRIPTION
**Pull request by barrutko**
The Receiver contract is still missing

### CODE REVIEW

File: src/testing/ZkEVMDomain.sol [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        if (log.topics[0] != BRIDGE_EVENT_TOPIC) return false;
+
+        (uint8 messageType, uint32 originNetwork,,,,,,) =
+            abi.decode(log.data, (uint8, uint32, address, uint32, address, uint256, bytes, uint32));
+        return
+            log.emitter == address(L1_MESSENGER) && messageType == 1 && (host ? originNetwork == 0 : originNetwork == 1);
+    }
+
+    function _claimMessage(Vm.Log memory log, bool host) internal {
+        require(_isBridgeMessageEvent(log, host), "ZkEVMDomain: !bridgeMessage");
```
Comment (hexonaut) [inline src/testing/ZkEVMDomain.sol:99]: This check is redundant.

File: src/testing/ZkEVMDomain.sol [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        vm.deal(address(L1_MESSENGER), prevBalance + msgValue);
+
+        // mock bridge callback
+        // ref: https://github.com/0xPolygonHermez/zkevm-contracts/blob/main/contracts/PolygonZkEVMBridge.sol#L455-L465
+        vm.prank(address(L1_MESSENGER));
+        (bool success,) = destinationAddress.call{value: msgValue}(
+            abi.encodeCall(IBridgeMessageReceiver.onMessageReceived, (originAddress, originNetwork, metadata))
+        );
+        if (!success) {
+            revert MessageFailed();
```
Comment (hexonaut) [inline src/testing/ZkEVMDomain.sol:124]: This should propagate the actual revert message similar to other domains.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Some minor changes. This also lacks the receiver authentication, which needs to be added in a subsequent PR.

**Review by hexonaut — APPROVED** [review]

File: src/XChainForwarders.sol [hunk 3]
```diff
@@ -152,4 +161,35 @@ library XChainForwarders {
...
+        ICrossDomainZkEVM(l1CrossDomain).bridgeMessage(
+            destinationNetworkId,
+            destinationAddress,
+            forceUpdateGlobalExitRoot,
+            metadata
+        );
+    }
+
+    function sendMessageZkEVM(
+        uint32 destinationNetworkId,
```
Comment (hexonaut) [inline src/XChainForwarders.sol:182]: This should be defaulted. I think it's `1` for the rollup.

File: src/XChainForwarders.sol [hunk 4]
```diff
@@ -152,4 +161,35 @@ library XChainForwarders {
...
+        address destinationAddress,
+        bool forceUpdateGlobalExitRoot,
```
Comment (hexonaut) [inline src/XChainForwarders.sol:184]: This should be defaulted. Figure out what the default value should be.

File: test/ZkEVMIntegration.sol [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        if (!success) {
+            assembly {
+                revert(add(ret, 0x20), mload(ret))
+            }
+        }
+    }
+
+}
+
+contract ZkEVMIntegrationTest is IntegrationBaseTest {
```
Comment (lucas-manuel) [inline test/ZkEVMIntegration.sol:21]: This filename needs `.t.sol` at end of it

File: test/ZkEVMIntegration.sol [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+        host.selectFork();
+
+        ZkevmMessageOrdering moHost = new ZkevmMessageOrdering();
+
+        zkevm.selectFork();
+
+        ZkevmMessageOrdering moZkevm = new ZkevmMessageOrdering();
+
+        // Queue up two more L1 -> L2 messages
```
Comment (lucas-manuel) [inline test/ZkEVMIntegration.sol:40]: L2 -> L1

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Checked forwarders and domain code against PR, LGTM, some changes

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Checked forwarders and domain code against PR, LGTM, some changes

**Review by hexonaut — APPROVED** [review]
LGTM now after Lucas's changes.

File: src/testing/ZkEVMDomain.sol [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity >=0.8.0;
+
+import { Vm } from "forge-std/Vm.sol";
+import { StdChains } from "forge-std/StdChains.sol";
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/testing/ZkEVMDomain.sol:5]: Resolving as will be changed in new PR

**Review by lucas-manuel — APPROVED** [review]
LGTM

**Review by hexonaut — APPROVED** [review]

**Pull request merged**